logstash-core/src/test/java/org/logstash/config/ir/graph/GraphTest.java

```diff
@@ -139,6 +143,29 @@ public void copyTest() throws InvalidIRException {
         assertTrue(rv.sourceComponentEquals(lv));
     }
 
+    @Test
+    public void uniqueHashingOfSimilarLeaves() throws InvalidIRException {
+        // the initial implementation didn't handle this well, so we'll leave it here as a tricky test
+        // During the construction of if conditions we'd have a partial graph with identical vertices
+        // in the eyes of the hash algorithm.
+
+        IfStatement imperative = DSL.iIf(
+                DSL.eTruthy(DSL.eValue("1")),
+                DSL.iPlugin(PluginDefinition.Type.FILTER, "drop"),
+                DSL.iIf(
+                        DSL.eTruthy(DSL.eValue("2")),
+                        DSL.iPlugin(PluginDefinition.Type.FILTER, "drop"),
+                        DSL.iIf(
+                                DSL.eTruthy(DSL.eValue("3")),
+                                DSL.iPlugin(PluginDefinition.Type.FILTER, "drop")
+                        )
+                )
+        );
+
+        Graph g = imperative.toGraph();
+        g.validate(); // Double validate just because that's the point of this test
+    }
```

Comment:
@ph well, there's no assert here because `validate`will throw an `InvalidIRException` if it fails.